cpp/src/io/parquet/parquet.hpp

```diff
@@ -118,6 +118,16 @@ struct LogicalType {
   BsonType BSON;
 };
 
+/** Empty struct to signal the order defined by the physical or logical type */
+struct TypeDefinedOrder {};
+
+/**
+ * Union to specify the order used for the min_value and max_value fields for a column.
+ */
+struct ColumnOrder {
+  std::optional<TypeDefinedOrder> TYPE_ORDER;
+};
+
```

Comment:
The parquet spec says unions are like enhanced enums. Many of the fields are just placeholders to signify which field is valid, but then the fields can have state, as with `LogicalType`. Right now, `LogicalType` is a struct of structs, with a struct field that's a struct of bools to indicate which field is valid. I'm proposing replacing that setup with a struct of optional structs, where only one field has a value.

Reply:
Thinking more about this, since this is supposed to be thought of as an enum, we could do something like:
```c++
struct TypeDefinedOrder {};

struct SomethingElse {
  int some_state;
};

struct ColumnOrder {
  enum class Type {
    UNDEFINED = 0,
    TYPE_ORDER = 1,
    SOMETHING_ELSE = 2
  };
  std::optional<TypeDefinedOrder> type_order;
  std::optional<SomethingElse> something_else;

  Type operator() const {
    if (type_order.has_value()) { return Type::TYPE_ORDER; }
    if (something_else.has_value()) { return Type::SOMETHING_ELSE; }
    return Type::UNDEFINED;
  }
};
```
Then the ColumnOrder could be used in a switch statement, and the case for `SOMETHING_ELSE` could access `column_order.something_else.value().some_state`.  Empty structs could be replaced with a `bool` I suppose (the logic for empty struct is already separate from the non-empty case). 

I would probably do it this way if I was writing a generic thrift parser, but for just parsing the parquet types, I think this may be overkill 😉